pkg/renderer/sgml/renderer.go

```diff
@@ -0,0 +1,360 @@
+package sgml
+
+import (
+	"bytes"
+	htmltemplate "html/template"
+	"io"
+	"strings"
+	texttemplate "text/template"
+
+	"github.com/bytesparadise/libasciidoc/pkg/configuration"
+	"github.com/bytesparadise/libasciidoc/pkg/types"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+// Renderer implements the backend render interface by using sgml.
+type Renderer interface {
+
+	// Render renders a document to the given output stream.
+	Render(ctx *Context, doc Document, output io.Writer) (Metadata, error)
+
+	// SetFunction sets the named function.
+	SetFunction(name string, fn interface{})
+
+	// Templates returns the Templates used by this Renderer.
+	// It cannot be altered on a given Renderer, since the old
+	// templates may have already been parsed.
+	Templates() Templates
+}
+
+func NewRenderer(t *Templates) Renderer {
+	sr := &sgmlRenderer{
+		functions: make(funcMap),
+		templates: t, // TODO: should we copy this?
+	}
+	// Establish some default function handlers.
```

Comment:
this could be set as 
```
sr := &sgmlRenderer{
	functions: map[string]interface{}{
		"render": sr.renderElements,
		...
	},
	templates: t,
}
```

Reply:
Agreed.  Will make this change.

Reply:
Can't *quite* do it that way, because I'm using  methods on the function being initialized.

But I *can* do:

sr.functions = funcMap{
   ...
}

So I'll do that.